File: src/tree/IsCompleteBinaryTree.java
package tree;

import java.util.ArrayDeque;
import java.util.Queue;

public class IsCompleteBinaryTree {
    public static boolean isCompleteBinaryTree(Node<?> root) {
        if (root == null) return true;

        Queue<Node<?>> queue = new ArrayDeque<>();
        queue.offer(root);

        boolean leafNull = false; // 没有页节点
        while (!queue.isEmpty()) {
            Node<?> node = queue.poll();
            Node<?> l = node.left;
            Node<?> r = node.right;

            if ((leafNull && (l != null || r != null)) || (l == null && r != null)) return false;

            if (l != null) queue.offer(l);
            if (r != null) queue.offer(r);

            if (l == null && r == null) leafNull = true;
        }
        return true;
    }

    public static boolean isComplete(Node<?> root) {
        return process(root).isCBT;
    }

    public static Info process(Node<?> node) {
        if (node == null) return new Info(true, true, 0);

        Info lInfo = process(node.left);
        Info rInfo = process(node.right);

        int height = Math.max(lInfo.height, rInfo.height) + 1;

        boolean isFull = lInfo.isFull && rInfo.isFull && lInfo.height == rInfo.height;
        boolean isCBT = isFull;

        if (lInfo.isCBT && rInfo.isCBT) {
            if ((lInfo.isFull && lInfo.height == rInfo.height)
                    || (rInfo.isFull && lInfo.height == rInfo.height + 1)
                    || (lInfo.isFull && rInfo.isFull && lInfo.height == rInfo.height + 1)) {
                isCBT = true;
            }
        }

        return new Info(isFull, isCBT, height);
    }

    static class Info {
        boolean isFull;
        boolean isCBT;
        int height;

        public Info(boolean isFull, boolean isCBT, int height) {
            this.isFull = isFull;
            this.isCBT = isCBT;
            this.height = height;
        }
    }

    /*public static boolean isComplete(Node<?> root) {
        return process(root).isCBT;
    }

    public static Info process(Node<?> node) {
        if (node == null) return new Info(true, true, 0);

        Info lInfo = process(node.left);
        Info rInfo = process(node.right);

        int height = Math.max(lInfo.height, rInfo.height) + 1;

        boolean isFull = lInfo.isFull && rInfo.isFull && lInfo.height == rInfo.height,
                isCBT = false;

        if (isFull) {
            isCBT = true;
        } else {
            if (lInfo.isCBT && rInfo.isCBT) {
                if ((lInfo.isFull && lInfo.height == rInfo.height)
                        || (rInfo.isFull && lInfo.height == rInfo.height + 1)
                        || (lInfo.isFull && rInfo.isFull && lInfo.height == rInfo.height + 1)) {
                    isCBT = true;
                }
            }
        }

        return new Info(isFull, isCBT, height);
    }

    static class Info {
        boolean isFull; // 是否是满二叉树 没有子节点
        boolean isCBT;
        int height;

        public Info(boolean isFull, boolean isCBT, int height) {
            this.isFull = isFull;
            this.isCBT = isCBT;
            this.height = height;
        }
    }*/
}

File: src/tree/is_complete_binary_tree.py
from collections import deque
from dataclasses import dataclass
from typing import Any, Optional


@dataclass
class Node:
    val: Any
    left: Optional["Node"] = None
    right: Optional["Node"] = None


def is_complete_binary_tree(root: Node) -> bool:
    """判断一个树是否是完全二叉树
    什么情况下不是完全二叉树?
        1. 左边没节点, 但是右边有
        2. 只要左右子节点其中一个没有, 那么接下来的节点还有子节点
    """
    q = deque()
    q.append(root)
    # 标记是否出现子节点不满的情况, 左右子节点其中一个出现过没有的情况(是否到达完全二叉树的结尾)
    flag = False
    while q:
        node = q.popleft()
        if (flag and (node.left or node.right)) or (not node.left and node.right):
            return False
        if node.left:
            q.append(node.left)
        if node.right:
            q.append(node.right)
        if not node.left or not node.right:
            flag = True
    return True


def is_complete_binary_tree_by_treedp_template(root: Node) -> bool:
    """判断一个树是否是完全二叉树, 树形dp套路版
    state info:
        x is cbt

    """

    @dataclass
    class Info:
        is_full: bool
        is_cbt: bool
        height: int

    def process(node):
        if not node:
            return Info(True, True, 0)
        linfo = process(node.left)
        rinfo = process(node.right)

        is_full = linfo.is_full and rinfo.is_full and linfo.height == rinfo.height
        height = max(linfo.height, rinfo.height) + 1
        is_cbt = is_full or (linfo.is_cbt and rinfo.is_full and linfo.height == rinfo.height + 1) or (
                linfo.is_full and rinfo.is_full and linfo.height == rinfo.height + 1) or (
                         linfo.is_full and rinfo.is_cbt and linfo.height == rinfo.height)

        return Info(is_full, is_cbt, height)

    return process(root).is_cbt
